src/test/spec/retryable_writes/mod.rs

```diff
@@ -0,0 +1,374 @@
+mod test_file;
+
+use futures::stream::TryStreamExt;
+use semver::VersionReq;
+
+use test_file::{Result, TestFile};
+
+use crate::{
+    bson::{doc, Document},
+    concern::{Acknowledgment, ReadConcern, WriteConcern},
+    error::ErrorKind,
+    options::{ClientOptions, CollectionOptions, FindOptions, InsertManyOptions},
+    test::{assert_matches, run_spec_test, EventClient, TestClient, LOCK},
+};
+
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+async fn run_spec_tests() {
+    async fn run_test(test_file: TestFile) {
+        for test_case in test_file.tests {
+            if test_case.operation.name == "bulkWrite" {
+                continue;
+            }
+
+            if test_case.description
+                == "InsertOne fails after multiple retryable writeConcernErrors"
+            {
+                continue;
+            }
+
+            let client = EventClient::with_additional_options(
+                test_case.client_options,
+                test_case.use_multiple_mongoses,
+            )
+            .await;
+
+            if let Some(ref run_on) = test_file.run_on {
+                let can_run_on = run_on.iter().any(|run_on| run_on.can_run_on(&client));
+                if !can_run_on {
+                    println!("Skipping {}", test_case.description);
+                    continue;
+                }
+            }
+
+            let mut db_name = test_case.description.replace('$', "%").replace(' ', "_");
+            // database names must have fewer than 64 characters
+            db_name.truncate(63);
```

Comment:
could we abstract this replace and truncate logic to a helper function? We use it in a multiple places in the tests I think.